CASSANDRA-16456 - Add Plugin Support for CQLSH

bhouse99 opened the pull request:
Added plugin support for CQLSH.  Support is available as part of Command line and via additions to the CQLSHRC. file.
For examples see CEP Documentation.

	* JIRA: https://issues.apache.org/jira/browse/CASSANDRA-16456
	* CEP: https://cwiki.apache.org/confluence/display/CASSANDRA/CEP-16%3A+Auth+Plugin+Support+for+CQLSH

bhouse99 committed:
CASSANDRA-16456 - Add Plugin Support for CQLSH

Added plugin support for CQLSH.  Support is available as part of Command line and via additions to the CQLSHRC. file.
For examples see CEP Documentation.

	* JIRA: https://issues.apache.org/jira/browse/CASSANDRA-16456
	* CEP: https://cwiki.apache.org/confluence/display/CASSANDRA/CEP-16%3A+Auth+Plugin+Support+for+CQLSH

smiklosovic committed:
added puresasl library

maulin-vasavada commented on bin/cqlsh.py:
```diff
@@ -2384,7 +2441,8 @@ def main(options, hostname, port):
                       connect_timeout=options.connect_timeout,
-                      encoding=options.encoding)
+                      encoding=options.encoding,
+                      auth_provider=load_auth_provider(options))
```
Can you please provide any test cases and samples for these changes?

bhouse99 replied:
Will add tests in the vein of sslhandling.

bhouse99 replied:
tests added

maulin-vasavada replied:
thanks!

maulin-vasavada commented on bin/cqlsh.py:
```diff
@@ -98,6 +99,33 @@
+    if not hasattr(passed_parser.values, 'auth_provider_extended') or parser.values.auth_provider_extended is None:
+        passed_parser.values.auth_provider_extended = {}
+
+    passed_parser.values.auth_provider_extended[key] = value
```
The name `auth_provider_extended` is little confusing in my opinion. Since it is set of properties that the pluggable AuthProvider can use, would `auth_provider_properties` or `auth_provider_config` make more sense?

bhouse99 replied:
Recently changed this to auth_provider_
config.

maulin-vasavada replied:
thanks!

maulin-vasavada commented on bin/cqlsh.py:
```diff
@@ -98,6 +99,33 @@
+
+def add_auth_provider_extended_options(passed_parser, argv):
+    for arg in argv:
+        if arg.startswith('--AuthProviderExtended'):
```
Similar to my previous comment, would it make more sense to name this `AuthProviderProperty`/`AuthProviderConfig` (I believe this is a single property that can be overridden by the command line)?

bhouse99 replied:
Renamed to auth_provider_config.

bhouse99 committed:
CASSANDRA-16456 - Add Plugin Support for CQLSH

Updated params to be snake case
AuthProvider to auth_provider.  AuthProviderExtendendProperties = auth_provider_extended_properties

	JIRA: https://issues.apache.org/jira/browse/CASSANDRA-16456

	CEP: https://cwiki.apache.org/confluence/display/CASSANDRA/CEP-16%3A+Auth+Plugin+Support+for+CQLSH

bhouse99 committed:
CASSANDRA-16456 - Add Plugin Support for CQLSH

Changed auth_provider_extended properties to auth_provider_config

bhouse99 committed:
Added library function for custom auth_provider loading

Added a library function for custom auth_provider loading
added tests and config to verify.

	JIRA: https://issues.apache.org/jira/browse/CASSANDRA-16456
	CEP: https://cwiki.apache.org/confluence/display/CASSANDRA/CEP-16%3A+Auth+Plugin+Support+for+CQLSH

bhouse99 committed:
Updated changes.txt

bhouse99 committed:
Minor clean up in removing unusued param

smiklosovic commented on CHANGES.txt:
```diff
@@ -24,6 +24,7 @@
  * Add diagnostic events for guardrails (CASSANDRA-17197)
  * Pre hashed passwords in CQL (CASSANDRA-17334)
  * Increase cqlsh version (CASSANDRA-17432)
+ * Added cqlsh custom auth provider support (CASSANDRA-16456)
```
@bhouse99 remove this line, I ll take care of that on the merge itself. No worries. We would have to rebase / fix it all over again as other people are adding stuff there in the meanwhile.

bhouse99 replied:
Kk, will do will avoid modifying in the future.

smiklosovic commented on bin/cqlsh.py:
```diff
@@ -405,7 +405,6 @@ def deserialize_date_fallback_int(byts, protocol_version):
     # Return cassandra.cqltypes.EMPTY instead of None for empty values
     cassandra.cqltypes.CassandraType.support_empty_values = True
 
-
```
should not be this kept?

smiklosovic commented on pylib/cqlshlib/authproviderhandling.py:
```diff
@@ -0,0 +1,79 @@
+            return None
+    
+    module_name = get_setting('auth_provider', 'module')
+    class_name = get_setting('auth_provider', 'classname')
```
what if there is no `class_name` but `module` is there? So `class_name` will be `None` but you are not checking it as you do upon `module_name`. Could you also add a test for that, please?

smiklosovic commented on pylib/cqlshlib/authproviderhandling.py:
```diff
@@ -0,0 +1,79 @@
+    Function which loads a custom auth provider from available config.
+
+    Params:
+    * config_file ..: path to cqlsh config file (usually ~/.cqlshrc).
```
the default path is actually `~/.cassandra/cqlshrc`

bhouse99 committed:
CASSANDRA-16456 - Add Plugin Support for CQLSH

updated as per CR from smiklosovic

bhouse99 committed:
CASSANDRA-16456 - Adds PlainTextAuthProvider Transparancy

Updated to add transparancy with the old way of using PlainTextAuthProvider and specifying PlainTextAuthProvider explicitly

AuthProviders now only read a single key [auth_provider] in the
cqlshrc file (instead of [auth_provider_config]).

Auth providers will now read properties stored under their
class name in credentials file.

	JIRA: https://issues.apache.org/jira/browse/CASSANDRA-16456
	CEP: https://cwiki.apache.org/confluence/display/CASSANDRA/CEP-16%3A+Auth+Plugin+Support+for+CQLSH

bhouse99 committed:
CASSANDRA-16456 - Pylinted new authproviderhandling

Made the new source file up-to-date with pylint.

   JIRA: https://issues.apache.org/jira/browse/CASSANDRA-16456
   CEP: https://cwiki.apache.org/confluence/display/CASSANDRA/CEP-16%3A+Auth+Plugin+Support+for+CQLSH

bhouse99 committed:
CASSANDRA-16456 - removes new ish plain_text_auth section

Removed the [plain_text_auth] section from credentials.  This is because
the new auth provider mechanics made this obsolete.

   JIRA: https://issues.apache.org/jira/browse/CASSANDRA-16456
   CEP: https://cwiki.apache.org/confluence/display/CASSANDRA/CEP-16%3A+Auth+Plugin+Support+for+CQLSH

smiklosovic commented on bin/cqlsh.py:
```diff
@@ -2091,7 +2095,6 @@ def is_file_secure(filename):
     # This is to allow "sudo cqlsh" to work with user owned credentials file.
     return (uid == 0 or st.st_uid == uid) and stat.S_IMODE(st.st_mode) & (stat.S_IRGRP | stat.S_IROTH) == 0
 
-
```
@bhouse99 this line should not be removed. I think there is 2 lines policy in Python between methods.

smiklosovic commented on bin/cqlsh.py:
```diff
@@ -2169,14 +2180,15 @@ def read_options(cmdlineargs, environment):
         rawcredentials.read(options.credentials)
 
         # handling password in the same way as username, priority cli > credentials > cqlshrc
         options.password = option_with_default(rawcredentials.get, 'plain_text_auth', 'password', password_from_cqlshrc)
```
@bhouse99 any reason why you have not removed this line, similarly as it is done for username case above?

smiklosovic commented on bin/cqlsh.py:
```diff
@@ -2169,14 +2180,15 @@ def read_options(cmdlineargs, environment):
 
         # use the username from credentials file but fallback to cqlshrc if username is absent from the command line parameters
-        options.username = option_with_default(credentials.get, 'plain_text_auth', 'username', username_from_cqlshrc)
+        options.username = username_from_cqlshrc
```
@bhouse99 I do not get this change. So we are populating `options.username` with `username_from_cqlshrc`. So for what reason do we read `credentials.read(options.credentials)` 3 lines above? That seems to me like dead code?

bhouse99 committed:
CASSANDRA-16456 - Passes username pass to all authproviders

Updated to always pass the username and password to auth providers
regardless of whether they are PlainTextAuthProvider.

added check for insecure credentials file

Defaults to PlainTextAuthProvider when not specified

	JIRA: https://issues.apache.org/jira/browse/CASSANDRA-16456

	CEP: https://cwiki.apache.org/confluence/display/CASSANDRA/CEP-16%3A+Auth+Plugin+Support+for+CQLSH

bhouse99 committed:
CASSANDRA-16456 - Minor updates from review

Updated to fix minor formating boo-boos, and remove
some dead code

Verified manually with sigv4 and plaintextauth
verified with cassandra cqlsh automated tests

	JIRA: https://issues.apache.org/jira/browse/CASSANDRA-16456

	CEP: https://cwiki.apache.org/confluence/display/CASSANDRA/CEP-16%3A+Auth+Plugin+Support+for+CQLSH

smiklosovic commented:
merged here https://github.com/apache/cassandra/commit/381c2a4fa82db80543fd67da6669de022d7576c4
